File: problems/分治/169.多数元素.py
#!/usr/bin/env python3
"""
169. 多数元素
"""
def majority_element(nums: list[int]) -> int:
    count = 0
    candidate = None
    for num in nums:
        if count == 0:
            candidate = num
        count += (1 if num == candidate else -1)
    return candidate

if __name__ == "__main__":
    # 测试用例
    print(majority_element([3, 2, 3]))  # 预期: 3
    print(majority_element([2, 2, 1, 1, 1, 2, 2]))  # 预期: 2
    print(majority_element([1]))  # 预期: 1

File: problems/分治/169.多数元素.cpp
#include <iostream>
#include <vector>
using namespace std;

int majorityElement(vector<int>& nums) {
    int count = 0, candidate = 0;
    for (int num : nums) {
        if (count == 0) candidate = num;
        count += (num == candidate ? 1 : -1);
    }
    return candidate;
}

int main() {
    // 测试用例
    vector<int> nums1 = {3, 2, 3};
    cout << majorityElement(nums1) << endl;  // 预期: 3
    vector<int> nums2 = {2, 2, 1, 1, 1, 2, 2};
    cout << majorityElement(nums2) << endl;  // 预期: 2
    vector<int> nums3 = {1};
    cout << majorityElement(nums3) << endl;  // 预期: 1
    return 0;
}
